rllib/agents/ppo/ppo.py

```diff
@@ -410,6 +418,86 @@ def get_default_policy_class(self, config: TrainerConfigDict) -> Type[Policy]:
+        self._counters[NUM_ENV_STEPS_SAMPLED] += num_env_steps
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += num_agent_steps
+
+        # Combine all batches at once and treat as multi-agent from here on.
+        train_batch = SampleBatch.concat_samples(sample_batches)
+        train_batch = train_batch.as_multi_agent()
+
+        # Standardize `advantages` values in train_batch.
+        for policy_id, batch in train_batch.policy_batches.items():
+            if Postprocessing.ADVANTAGES not in batch:
+                raise KeyError(
+                    f"`{Postprocessing.ADVANTAGES}` not found in SampleBatch for "
+                    f"policy `{policy_id}`! Maybe this policy fails to add "
+                    f"`{Postprocessing.ADVANTAGES}` in its `postprocess_trajectory` "
+                    f"method? Or this policy is not meant to learn at all and you "
+                    "forgot to remove it from the list under `config."
+                    "multiagent.policies_to_train`."
+                )
+            batch[Postprocessing.ADVANTAGES] = standardized(
+                batch[Postprocessing.ADVANTAGES])
+
+        # Use simple optimizer (only for multi-agent or tf-eager; all other
+        # cases should use the multi-GPU optimizer, even if only using 1 GPU).
+        if self.config.get("simple_optimizer") is True:
+            train_results = train_one_step(self, train_batch)
+        else:
+            train_results = multi_gpu_train_one_step(self, train_batch)
+
+        global_vars = {
+            "timestep": self._counters[NUM_AGENT_STEPS_SAMPLED],
+        }
+
+        # Update weights - after learning on the local worker - on all remote
+        # workers.
+        if self.workers.remote_workers():
+            with self._timers[WORKER_UPDATE_TIMER]:
+                self.workers.sync_weights(global_vars=global_vars)
+
+        # Update KL after each round of training.
+        def update(policy, policy_id):
+            assert LEARNER_STATS_KEY not in train_results, (
+                "{} should be nested under policy id key".format(LEARNER_STATS_KEY),
+                train_results,
+            )
+            if policy_id in train_results:
+                kl = train_results[policy_id][LEARNER_STATS_KEY].get("kl")
+                assert kl is not None, (train_results, policy_id)
+                # Make the actual `Policy.update_kl()` call.
+                policy.update_kl(kl)
+            else:
+                logger.warning("No data for {}, not updating kl".format(policy_id))
+
+        # Update KL on all trainable policies within the local (trainer)
+        # Worker.
+        self.workers.local_worker().foreach_policy_to_train(update)
+
+        # Update global vars on .
+        self.workers.local_worker().set_global_vars(global_vars)
+
+        # Warn about bad reward scales and return training metrics.
+        #warn_about_bad_reward_scales(self.config, train_results)
```

Comment:
delete? why have commented out lines here?

Reply:
Yeah, we got to fix this first, sorry.
Metrics from rollouts are pulled after(!) returning from `training_iteration`, so having this here won't work. We could add the check at the beginning of the iteration fn and run the analysis (only needs episode_reward_mean and vf_loss) on the previous iteration's data. ... Let me fix this.
I also noticed that we don't perform these checks in the multi-agent case, so we can add this here as well.